From cd8ec59e734883b88737a1c05fc9339d093b25fb Mon Sep 17 00:00:00 2001
Message-Id: <cd8ec59e734883b88737a1c05fc9339d093b25fb.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: vkorjani <vikas.korjani@intel.com>
Date: Thu, 31 Jul 2014 12:29:41 +0530
Subject: [PATCH 164/312] FOR_UPSTREAM [VPG]: drm/i915: Loading panel width
 and height from bios

Getting the width and height of the panel from the VBT.

Issue : GMIN-674
Change-Id: Iee014501be24ba13c051a35371463a333fe38adf
Signed-off-by: vkorjani <vikas.korjani@intel.com>
Signed-off-by: Deepak M <m.deepak@intel.com>
---
 drivers/gpu/drm/i915/intel_bios.c          |    4 ++++
 drivers/gpu/drm/i915/intel_dsi.c           |    4 ++++
 drivers/gpu/drm/i915/intel_dsi.h           |    2 ++
 drivers/gpu/drm/i915/intel_dsi_panel_vbt.c |   18 ++++++++++++++++++
 4 files changed, 28 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_bios.c b/drivers/gpu/drm/i915/intel_bios.c
index ce7d484..fa9321c 100644
--- a/drivers/gpu/drm/i915/intel_bios.c
+++ b/drivers/gpu/drm/i915/intel_bios.c
@@ -101,6 +101,10 @@ fill_detail_timing_data(struct drm_display_mode *panel_fixed_mode,
 		((dvo_timing->vblank_hi << 8) | dvo_timing->vblank_lo);
 	panel_fixed_mode->clock = dvo_timing->clock * 10;
 	panel_fixed_mode->type = DRM_MODE_TYPE_PREFERRED;
+	panel_fixed_mode->width_mm = dvo_timing->h_image |
+		((dvo_timing->max_hv & 15) << 8);
+	panel_fixed_mode->height_mm = dvo_timing->v_image |
+		((dvo_timing->max_hv >> 4) << 8);
 
 	if (dvo_timing->hsync_positive)
 		panel_fixed_mode->flags |= DRM_MODE_FLAG_PHSYNC;
diff --git a/drivers/gpu/drm/i915/intel_dsi.c b/drivers/gpu/drm/i915/intel_dsi.c
index 7c473b0..3ae49e5 100644
--- a/drivers/gpu/drm/i915/intel_dsi.c
+++ b/drivers/gpu/drm/i915/intel_dsi.c
@@ -659,6 +659,7 @@ intel_dsi_detect(struct drm_connector *connector, bool force)
 static int intel_dsi_get_modes(struct drm_connector *connector)
 {
 	struct intel_connector *intel_connector = to_intel_connector(connector);
+	struct intel_dsi *intel_dsi = intel_attached_dsi(connector);
 	struct drm_display_mode *mode;
 
 	DRM_DEBUG_KMS("\n");
@@ -676,6 +677,9 @@ static int intel_dsi_get_modes(struct drm_connector *connector)
 	}
 
 	drm_mode_probed_add(connector, mode);
+
+	/*Fill the panel info here*/
+	intel_dsi->dev.dev_ops->get_info(0, connector);
 	return 1;
 }
 
diff --git a/drivers/gpu/drm/i915/intel_dsi.h b/drivers/gpu/drm/i915/intel_dsi.h
index 447a07f..003bda2 100644
--- a/drivers/gpu/drm/i915/intel_dsi.h
+++ b/drivers/gpu/drm/i915/intel_dsi.h
@@ -70,6 +70,8 @@ struct intel_dsi_device {
 struct intel_dsi_dev_ops {
 	bool (*init)(struct intel_dsi_device *dsi);
 
+	void (*get_info)(int pipe, struct drm_connector *connector);
+
 	void (*panel_reset)(struct intel_dsi_device *dsi);
 
 	void (*disable_panel_power)(struct intel_dsi_device *dsi);
diff --git a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
index dc5ab89..cd0985b 100644
--- a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
+++ b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
@@ -231,6 +231,23 @@ static void generic_exec_sequence(struct intel_dsi *intel_dsi, char *sequence)
 	}
 }
 
+static void generic_get_panel_info(int pipe, struct drm_connector *connector)
+{
+	struct intel_connector *intel_connector = to_intel_connector(connector);
+	DRM_DEBUG_KMS("\n");
+	if (!connector)
+		return;
+
+	if (pipe == 0) {
+		connector->display_info.width_mm =
+			intel_connector->panel.fixed_mode->width_mm;
+		connector->display_info.height_mm =
+			intel_connector->panel.fixed_mode->height_mm;
+	}
+
+	return;
+}
+
 static bool generic_init(struct intel_dsi_device *dsi)
 {
 	struct intel_dsi *intel_dsi = container_of(dsi, struct intel_dsi, dev);
@@ -550,6 +567,7 @@ static void generic_destroy(struct intel_dsi_device *dsi) { }
 /* Callbacks. We might not need them all. */
 struct intel_dsi_dev_ops vbt_generic_dsi_display_ops = {
 	.init = generic_init,
+	.get_info = generic_get_panel_info,
 	.mode_valid = generic_mode_valid,
 	.mode_fixup = generic_mode_fixup,
 	.panel_reset = generic_panel_reset,
-- 
1.7.9.5

